### PR TITLE
Introduce AttributedText

### PR DESCRIPTION
**Pull request by kylebshr**
`AttributedText` allows you to apply strongly-typed attributes to strings (much like the `AttributedString` type introduced in iOS 15). You can then access the `attributedString` property to get an attributed string with those attributes applied.

For example:

```swift
var text = AttributedText(string: "Hello, world")
// Apply a font to the entire range
text.font = .systemFont(ofSize: 20)

// Apply a color to part of the string
let range = text.string.range(of: "world")!
text[range].color = .blue

// Render the attributed text
let label = AttributedLabel(attributedText: text.attributedString)
```

### CODE REVIEW

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+/// ```
+///
+@dynamicMemberLookup public struct AttributedText {
+
+    /// The wrapped string, with no attributes.
+    public let string: String
+
+    /// An `NSAttributedString` representation of the attributed text.
+    public var attributedString: NSAttributedString {
+        mutableAttributedString
```
Comment (kyleve) [inline BlueprintUI/Sources/AttributedText/AttributedText.swift:29]: Nit: Return a `.copy()` of this – otherwise you're returning the actual underlying mutable value, which can continue to mutate.

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    /// range. (Setting an attribute will set it across the subscript range regardless of any existing contents).
+    ///
+    public subscript(range: Range<String.Index>) -> TextAttributeContainer {
+        get {
+            let range = NSRange(range, in: string)
+            return makeAttributeStore(range: range)
+        }
+        set {
+            let range = NSRange(range, in: string)
+            mutableAttributedString.addAttributes(newValue.storage, range: range)
```
Comment (kyleve) [inline BlueprintUI/Sources/AttributedText/AttributedText.swift:88]: `NSMutableAttributedString` is a mutable reference (class) type. This struct is a class type; you need to implement copy on write semantics here otherwise each copy of the struct is gonna reference the same underlying mutable string

File: BlueprintUI/Sources/AttributedText/AttributedTextKey.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        get { self[ColorKey.self] }
+        set { self[ColorKey.self] = newValue }
+    }
+}
+
+// MARK: Kern
+
+public enum KernKey: AttributedTextKey {
+    public typealias Value = CGFloat
+    public static var name: NSAttributedString.Key { .kern }
```
Comment (watt) [inline BlueprintUI/Sources/AttributedText/AttributedTextKey.swift:49]: For the built-in keys, I think we might want to expose tracking instead of kerning, via `kCTTrackingAttributeName as NSAttributedString.Key`.

Check out [this comment thread where we added it to POSUI for context](https://github.com/squareup/ios-register/pull/49844#discussion_r554402886).

File: BlueprintUI/Sources/AttributedText/AttributedText.swift (hunk 2 above)
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:88]: Ah yep, very good catch...

Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift]: Added a test for this and pushed up a fix

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    private var mutableAttributedString: NSMutableAttributedString
+
+    /// Create some `AttributedText` from a plain string.
+    public init(string: String) {
+        self.string = string
+        self.mutableAttributedString = NSMutableAttributedString(string: string)
+    }
+
+    /// Create some `AttributedText` from an attributed string. The attributes are preserved.
+    public init(attributedString: NSAttributedString) {
```
Comment (kylebshr) [inline BlueprintUI/Sources/AttributedText/AttributedText.swift:41]: Should this also just be `string:`? Should both inits have no name for the parameter?

```suggestion
    public init(string: NSAttributedString) {
```

File: BlueprintUI/Tests/AttributedTextTests.swift [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        let hello = text.range(of: "Hello")!
+        let world = text.range(of: "world")!
+        let loWo = text.range(of: "lo wo")!
+        let rld = text.range(of: "rld")!
+
+        text[hello].font = font
+        text[world].color = green
+        text[loWo].color = blue
+
+        XCTAssertEqual(text[hello].font, font, "The font should be returned when it applies to the whole range")
```
Comment (watt) [inline BlueprintUI/Tests/AttributedTextTests.swift:40]: Can you also test that a contained range like `"ell"` and a containing range like `"hello wo"` are both nil?

**Review by watt — APPROVED** [review]

**Comment by watt** [comment]
For anybody else who got curious about how hard a custom storage implementation would be:
- [CFAttributedString](https://github.com/apple/swift-corelibs-foundation/blob/main/CoreFoundation/String.subproj/CFAttributedString.c)
- [the new AttributedString](https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/AttributedString/AttributedString.swift)

I was expecting something like an [interval tree](https://en.wikipedia.org/wiki/Interval_tree), but both of them seem to boil down to arrays of runs.

File: BlueprintUI/Sources/AttributedText/AttributedText.swift (hunk 4 above)
Comment (n8chur) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:41]: I'm pretty sure we want to omit the parameter here based on the [Swift API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/#type-conversion), but I might be misunderstanding them.

Comment (n8chur) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:41]: Or maybe we are supposed to put the parameter for `String` because it's "narrowing" and not put the parameter for `NSAttributedString` because it's just converting?

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+/// text[range].color = .blue
+///
+/// // Render the attributed text
+/// let label = AttributedLabel(attributedText: text.attributedString)
+/// ```
+///
+@dynamicMemberLookup public struct AttributedText {
+
+    /// The wrapped string, with no attributes.
+    public let string: String
```
[the remark this answers is not in the document]
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:25]: Hmm I guess so, initially I didn't expose `range(of:)` so consumers could use this string directly. Maybe it doesn't need to be public yet?

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    /// for specific ranges using strong types:
+    ///
+    /// ```swift
+    /// var text = AttributedText(string: "Hello, world")
+    /// let range = text.string.range(of: "Hello")!
+    /// text[range].font = .systemFont(ofSize: 20)
+    /// ```
+    ///
+    /// Note that the returned `TextAttributeContainer` will only contain attributes that apply to the entire subscript
+    /// range. (Setting an attribute will set it across the subscript range regardless of any existing contents).
```
[the remark this answers is not in the document]
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:79]: If the attributes extend beyond the range they'll still be returned! It's only attributes that apply to part of the range that aren't returned

File: BlueprintUI/Sources/AttributedText/AttributedText.swift (hunk 2 above)
Comment (kyleve) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift]: Nice

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    ///
+    public subscript(range: Range<String.Index>) -> TextAttributeContainer {
```
[the remark this answers is not in the document]
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:81]: Yep, that works!

File: BlueprintUI/Tests/AttributedTextTests.swift (hunk 5 above)
Comment (kylebshr) [reply BlueprintUI/Tests/AttributedTextTests.swift:46]: Done (although "ell" returns the font)

File: BlueprintUI/Sources/AttributedText/AttributedText.swift (hunk 4 above)
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:41]: I'm for that

File: BlueprintUI/Sources/AttributedText/AttributedTextKey.swift (hunk 3 above)
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedTextKey.swift]: Done

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        for attribute in attributesInRange {
+            mutableAttributedString.enumerateAttribute(
+                attribute,
+                in: range,
+                options: []
+            ) { value, attributeRange, _ in
+                if let value = value, attributeRange == range {
+                    store.storage[attribute] = value
+                }
+            }
```
Comment (watt) [inline BlueprintUI/Sources/AttributedText/AttributedText.swift:153]: So if I'm following what you're saying here:

> If the attributes extend beyond the range they'll still be returned! It's only attributes that apply to part of the range that aren't returned

The `attributedRange` passed into this block by `enumerateAttribute` will be no larger than `range`, even if that attributes's range does extend outside of `range`?

File: BlueprintUI/Tests/AttributedTextTests.swift [hunk 10]
```diff
@@ -0,0 +1,129 @@
+import XCTest
+@testable import BlueprintUI
+
+class AttributedTextTests: XCTestCase {
```
Comment (kyleve) [inline BlueprintUI/Tests/AttributedTextTests.swift:4]: Nit: It would also be good to add some test using emojis and non-ascii characters to validate this does what we expect there as well.

Comment (kylebshr) [reply BlueprintUI/Tests/AttributedTextTests.swift:4]: Done

File: BlueprintUI/Sources/AttributedText/AttributedText.swift (hunk 9 above)
Comment (kylebshr) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:162]: Yes! Although it's not documented well, that's the behavior I'm seeing (and expected, tbh)

**Review by kyleve — APPROVED** [review]

**Review by watt — APPROVED** [review]

**Review by n8chur — APPROVED** [review]
Nice work!

File: BlueprintUI/Sources/AttributedText/AttributedText.swift [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+
+    private var entireRange: Range<String.Index> {
+        string.startIndex..<string.endIndex
+    }
+
+    private mutating func addAttributes(attributes: TextAttributeContainer, to range: NSRange) {
+        if !isKnownUniquelyReferenced(&mutableAttributedString) {
+            mutableAttributedString = NSMutableAttributedString(attributedString: mutableAttributedString)
```
[the remark this answers is not in the document]
Comment (kyleve) [reply BlueprintUI/Sources/AttributedText/AttributedText.swift:128]: Shouldn't matter beyond being a bit less wordy.
